pinot-connectors/pinot-flink-connector/src/main/java/org/apache/pinot/connector/flink/sink/FlinkSegmentWriter.java

```diff
@@ -0,0 +1,266 @@
+   * segment and not available in the buffer anymore.
+   *
+   * <p>Successful completion of segment will return the segment URI, and the URI includes a
+   * sequence id indicating the part number. The sequence id is initialized to 0 and each successful
+   * flush will increment the sequence id by 1. The buffer will be reset and ready to accept further
+   * records via <code>collect()</code> If an exception is thrown, the buffer will not be reset and
+   * so, <code>flush()</code> can be invoked repeatedly in a retry loop. If a successful invocation
+   * is not achieved,<code>close()</code> followed by <code>init </code> will have to be called in
+   * order to reset the buffer and resume record writing.
+   *
+   * @return URI of the generated segment
+   * @throws IOException
+   */
+  @Override
+  public URI flush()
+      throws IOException {
+
+    LOGGER.info("Beginning flush for Pinot table: {} with {} records", _tableNameWithType, _rowCount);
+    _recordWriter.close();
+
+    // Create temp dir for flush
+    File flushDir = new File(_stagingDir, "flush_dir_" + System.currentTimeMillis());
+    Preconditions.checkState(flushDir.mkdirs(), "Failed to create flush dir: %s", flushDir);
+
+    try {
+      // Segment dir
+      File segmentDir = new File(flushDir, "segment_dir");
+
+      // Make BatchIngestionConfig for flush
+      Map<String, String> batchConfigMapOverride = new HashMap<>(_batchConfig.getBatchConfigMap());
+      batchConfigMapOverride.put(BatchConfigProperties.INPUT_DIR_URI, _bufferFile.getAbsolutePath());
+      batchConfigMapOverride.put(BatchConfigProperties.OUTPUT_DIR_URI, segmentDir.getAbsolutePath());
+      batchConfigMapOverride.put(BatchConfigProperties.INPUT_FORMAT, BUFFER_FILE_FORMAT.toString());
+      BatchIngestionConfig batchIngestionConfig = new BatchIngestionConfig(Lists.newArrayList(batchConfigMapOverride),
+          _batchIngestionConfig.getSegmentIngestionType(), _batchIngestionConfig.getSegmentIngestionFrequency());
+
+      // Build segment
+      SegmentGeneratorConfig segmentGeneratorConfig =
+          IngestionUtils.generateSegmentGeneratorConfig(_tableConfig, _schema, batchIngestionConfig);
+      segmentGeneratorConfig.setSequenceId(_seqId);
+      String segmentName = IngestionUtils.buildSegment(segmentGeneratorConfig);
+      LOGGER.info("Successfully built segment: {} of sequence {} for Pinot table: {}", segmentName, _seqId,
+          _tableNameWithType);
+
+      // Tar segment
+      File segmentTarFile =
+          new File(_outputDirURI, String.format("%s_%d%s", segmentName, _indexOfSubtask, Constants.TAR_GZ_FILE_EXT));
+      if (!_batchConfig.isOverwriteOutput() && segmentTarFile.exists()) {
+        segmentTarFile = new File(_outputDirURI,
+            String.format("%s_%d_%d%s", segmentName, _indexOfSubtask, System.currentTimeMillis(),
+                Constants.TAR_GZ_FILE_EXT));
+      }
```

Comment:
if batchConfig.isOverwriteOutput and segmentTarFile.exists, delete the existing file ? Or is the tar creation guaranteed to not get interfered by existence of the file already?

Reply:
yes. it guarantees. Each segment in each run will have its own file name, and the file will be deleted after the run